content/posts/0.28-release.md

```diff
@@ -0,0 +1,254 @@
+load dynamic libraries eagerly rather than lazily.
+
+We added runtime path to the WebAssembly specification for the shared library
+format, to the llvm WebAssembly object parser and linker, and to the `emcc`
+linker. We also updated the dynamic loader to use this information. Finally, we
+updated `pyodide-build` to emit the new runtime path data and we removed our
+patch to the dynamic loader.
+
+### Increased Adoption of JavaScript Promise Integration
+
+JavaScript Promise Integration (JSPI) officially became a Stage 4 finished
+proposal on April 8, 2025, and Chrome 137 (released May 27, 2025) now supports
+JSPI by default, no experimental flag needed.
+
+Pyodide has been a long-time experimenter with JSPI. We turned on several key
+JSPI features by default in Pyodide 0.27.7. This means you can now use
+`asyncio.run()` and `loop.run_until_complete()` in Pyodide to execute Python
+code in the browser to block for asynchronous operations. Previously, this
+capability was gated by the `enableRunUntilComplete` flag in `loadPyodide()`.
+Now, if your browser supports JSPI, these features are enabled automatically.
+See [our blog post about JSPI](https://blog.pyodide.org/posts/jspi/) for more
+information.
+
+### Support for `null` in the Python/JavaScript Foreign Function Interface
+
+As much as possible, the Pyodide foreign function interface tries to ensure that
+values round trip: if a JavaScript value is passed to Python and then back to
+JavaScript, it should come back `===` to the original value. Previously, both
+`null` and `undefined` were converted to `None` and `None` was converted to
+`undefined` so `null` would not round trip. This made it impossible to use
+JavaScript APIs that treat `null` and `undefined` differently.
+
+Fixing this was more difficult than it would seem. When JavaScript values are
+passed into C, we represent them as a WebAssembly `externref`. We want to use a
+special value to indicate that an error occured. There is a special WebAssembly
+instruction called `ref.is_null` to check whether an `externref` is `null`. All
+other `externrefs` were opaque to WebAssembly, and to find out about their
+identity we needed to call out to JavaScript. Calling out to JavaScript to test
+for an error is slow (we measured a 2-3% performance hit for this) so we used
+`null` to signal an error internally.
+
+In order to prevent a `null` that came from the user from being misintepreted as
+an error signal, we converted all `null` to `undefined` before passing them into
+C. This made supporting `null` in the FFI impossible.
+
+However, the new WebAssembly Garbage Collection (wasm-gc) feature adds new
+instructions to create and check for special externrefs that are not `null`. We
+switched to using this whenever it is supported. If wasm-gc is not supported, we
+fall back to using a JavaScript function to test for an error and put up with
+the 2-3% performance hit. Since December 2024, all major JavaScript runtimes
+support wasm-gc.
```

Comment:
```suggestion
However, the new WebAssembly Garbage Collection (wasm-gc) feature adds new
instructions to create and check for special externrefs that are not `null`. We
switched to using this whenever it is supported. If wasm-gc is not supported, we
fall back to using a JavaScript function to test for an error and put up with
the 2-3% performance hit. [Since December 2024, all major JavaScript runtimes
support wasm-gc](https://webassembly.org/features/).
```